pcs/pcs.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH PCS "8" "August 2013" "pcs 0.9.132" "System Administration Utilities"
.SH NAME
pcs \- pacemaker/corosync configuration system
.SH SYNOPSIS
.B pcs
[\fI\-f file\fR] [\fI\-h\fR] [\fIcommands\fR]...
.SH DESCRIPTION
Control and configure pacemaker and corosync.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Display usage and exit
.TP
\fB\-f\fR file
Perform actions on file instead of active CIB
.TP
\fB\-\-debug\fR
Print all network traffic and external commands run
.TP
\fB\-\-version\fR
Print pcs version information
.SS "Commands:"
.TP
cluster
Configure cluster options and nodes
.TP
resource
Manage cluster resources
.TP
stonith
Configure fence devices
.TP
constraint
Set resource constraints
.TP
property
Set pacemaker properties
.TP
acl
Set pacemaker access control lists
.TP
status
View cluster status
.TP
config
View and manage cluster configuration
.SS "resource"
.TP
show [resource id] [\fB\-\-full\fR] [\fB\-\-groups\fR]
Show all currently configured resources or if a resource is specified show the options for the configured resource.  If \fB\-\-full\fR is specified all configured resource options will be displayed.  If \fB\-\-groups\fR is specified, only show groups (and their resources).
.TP
list [<standard|provider|type>] [\fB\-\-nodesc\fR]
Show list of all available resources, optionally filtered by specified type, standard or provider. If \fB\-\-nodesc\fR is used then descriptions of resources are not printed.
.TP
describe <standard:provider:type|type>
Show options for the specified resource
.TP
create <resource id> <standard:provider:type|type> [resource options] [op <operation action> <operation options> [<operation action> <operation options>]...] [meta <meta options>...] [\fB\-\-clone\fR <clone options> | \fB\-\-master\fR <master options> | \fB\-\-group\fR <group name> [\fB\-\-before\fR <resource id> | \fB\-\-after\fR <resource id>]] [\fB\-\-disabled\fR]
Create specified resource.  If \fB\-\-clone\fR is used a clone resource is created if \fB\-\-master\fR is specified a master/slave resource is created.  If \fB\-\-group\fR is specified the resource is added to the group named.  You can use \fB\-\-before\fR or \fB\-\-after\fR to specify the position of the added resource relatively to some resource already existing in the group.  If \fB\-\-disabled\fR is specified the resource is not started automatically. Example: pcs resource create VirtualIP ocf:heartbeat:IPaddr2 \e ip=192.168.0.99 cidr_netmask=32 nic=eth2 op monitor interval=30s \e Create a new resource called 'VirtualIP' with IP address 192.168.0.99, netmask of 32, monitored everything 30 seconds, on eth2.
.TP
delete <resource id|group id|master id|clone id>
Deletes the resource, group, master or clone (and all resources within the group/master/clone).
.TP
enable <resource id> [\fB\-\-wait\fR[=n]]
Allow the cluster to start the resource. Depending on the rest of the configuration (constraints, options, failures, etc), the resource may remain stopped.  If \fB\-\-wait\fR is specified, pcs will wait up to 30 seconds (or 'n' seconds) for the resource to start and then return 0 if the resource is started, or 1 if the resource has not yet started.
.TP
disable <resource id> [\fB\-\-wait\fR[=n]]
Attempt to stop the resource if it is running and forbid the cluster from starting it again.  Depending on the rest of the configuration (constraints, options, failures, etc), the resource may remain started.  If \fB\-\-wait\fR is specified, pcs will wait up to 30 seconds (or 'n' seconds) for the resource to stop and then return 0 if the resource is stopped or 1 if the resource has not stopped.
.TP
debug\-start <resource id> [\fB\-\-full\fR]
This command will force the specified resource to start on this node ignoring the cluster recommendations and print the output from starting the resource. Using \fB\-\-full\fR will give more detailed output. This is mainly used for debugging resources that fail to start.
.TP
move <resource id> [destination node] [\fB\-\-master\fR] [lifetime=<lifetime>]
Move resource off current node (and optionally onto destination node). If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id). If lifetime is not specified it defaults to infinite.
.TP
ban <resource id> [node] [\fB\-\-master\fR] [lifetime=<lifetime>]
Prevent the resource id specified from running on the node (or on the current node it is running on if no node is specified). If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id). If lifetime is not specified it defaults to infinite.
.TP
clear <resource id> [node] [\fB\-\-master\fR]
Remove constraints created by move and/or ban on the specified resource (and node if specified). If \fB\-\-master\fR is used the scope of the command is limited to the master role and you must use the master id (instead of the resource id).
.TP
standards
List available resource agent standards supported by this installation. (OCF, LSB, etc.)
.TP
providers
List available OCF resource agent providers
.TP
agents [standard[:provider]]
List available agents optionally filtered by standard and provider
.TP
update <resource id> [resource options] [op [<operation action> <operation options>]...] [meta <meta operations>...]
Add/Change options to specified resource, clone or multi\-state resource.  If an operation (op) is specified it will update the first found operation with the same action on the specified resource, if no operation with that action exists then a new operation will be created (WARNING: all current options on the update op will be reset if not specified). If you want to create multiple monitor operations you should use the add_operation & remove_operation commands.
.TP
op add <resource id> <operation action> [operation properties]
Add operation for specified resource
.TP
op remove <resource id> <operation action> [<operation properties>...]
Remove specified operation (note: you must specify the exact operation properties to properly remove an existing operation).
.TP
op remove <operation id>
Remove the specified operation id
.TP
op defaults [options]
Set default values for operations, if no options are passed, lists currently configured defaults
.TP
meta <resource id | group id | master id | clone id> <meta options>
Add specified options to the specified resource, group, master/slave or clone.  Meta options should be in the format of name=value, options may be removed by setting an option without a value. Example: pcs resource meta TestResource failure\-timeout=50 stickiness=
.TP
group add <group name> <resource id> [resource id] ... [resource id] [\fB\-\-before\fR <resource id> | \fB\-\-after\fR <resource id>]
Add the specified resource to the group, creating the group if it does not exist.  If the resource is present in another group it is moved to the new group.  You can use \fB\-\-before\fR or \fB\-\-after\fR to specify the position of the added resources relatively to some resource already existing in the group.
.TP
group remove <group name> <resource id> [resource id] ... [resource id]
Remove the specified resource(s) from the group, removing the group if it no resources remain.
.TP
ungroup <group name> [resource id] ... [resource id]
Remove the group (Note: this does not remove any resources from the cluster) or if resources are specified, remove the specified resources from the group
.TP
clone <resource id | group id> [clone options]...
Setup up the specified resource or group as a clone
.TP
unclone <resource id | group name>
Remove the clone which contains the specified group or resource (the resource or group will not be removed)
.TP
master [<master/slave name>] <resource id | group name> [options]
Configure a resource or group as a multi\-state (master/slave) resource. Note: to remove a master you must remove the resource/group it contains.
.TP
manage <resource id> ... [resource n]
Set resources listed to managed mode (default)
.TP
unmanage <resource id> ... [resource n]
Set resources listed to unmanaged mode
.TP
defaults [options]
Set default values for resources, if no options are passed, lists currently configured defaults
.TP
cleanup [<resource id>]
Cleans up the resource in the lrmd (useful to reset the resource status and failcount). This tells the cluster to forget the operation history of a resource and re-detect its current state. This can be useful to purge knowledge of past failures that have since been resolved. If resource id is not specified then all resources/stonith devices will be cleaned up.
.TP
failcount show <resource id> [node]
Show current failcount for specified resource from all nodes or only on specified node
.TP
failcount reset <resource id> [node]
Reset failcount for specified resource on all nodes or only on specified node. This tells the cluster to forget how many times a resource has failed in the past.  This may allow the resource to be started or moved to a more preferred location.
.SS "cluster"
.TP
auth [node] [...] [\fB\-u\fR username] [\fB\-p\fR password] [\fB\-\-local\fR] [\fB\-\-force\fR]
Authenticate pcs to pcsd on nodes specified, or on all nodes configured in corosync.conf if no nodes are specified (authorization tokens are stored in ~/.pcs/tokens or /var/lib/pcsd/tokens for root).  By default all nodes are also authenticated to each other, using \fB\-\-local\fR only authenticates the local node (and does not authenticate the remote nodes with each other).  Using \fB\-\-force\fR forces re-authentication to occur.
.TP
setup [\fB\-\-start\fR] [\fB\-\-local\fR] [\fB\-\-enable\fR] \fB\-\-name\fR <cluster name> <node1[,node1-altaddr]> [node2[,node2-altaddr]] [..] [\fB\-\-transport\fR <udpu|udp>] [\fB\-\-rrpmode\fR active|passive] [\fB\-\-addr0\fR <addr/net> [[[\fB\-\-mcast0\fR <address>] [\fB\-\-mcastport0\fR <port>] [\fB\-\-ttl0\fR <ttl>]] | [\fB\-\-broadcast0\fR]] [\fB\-\-addr1\fR <addr/net> [[[\fB\-\-mcast1\fR <address>] [\fB\-\-mcastport1\fR <port>] [\fB\-\-ttl1\fR <ttl>]] | [\fB\-\-broadcast1\fR]]]] [\fB\-\-wait_for_all\fR=<0|1>] [\fB\-\-auto_tie_breaker\fR=<0|1>] [\fB\-\-last_man_standing\fR=<0|1> [\fB\-\-last_man_standing_window\fR=<time in ms>]] [\fB\-\-token\fR <timeout>] [\fB\-\-token_coefficient\fR <timeout>] [\fB\-\-join\fR <timeout>] [\fB\-\-consensus\fR <timeout>] [\fB\-\-miss_count_const\fR <count>] [\fB\-\-fail_recv_const\fR <failures>]
Configure corosync and sync configuration out to listed nodes. \fB\-\-local\fR will only perform changes on the local node, \fB\-\-start\fR will also start the cluster on the specified nodes, \fB\-\-enable\fR will enable corosync and pacemaker on node startup, \fB\-\-transport\fR allows specification of corosync transport (default: udpu), \fB\-\-rrpmode\fR allows you to set the RRP mode of the system. Currently only 'passive' is supported or tested (using 'active' is not recommended). The \fB\-\-wait_for_all\fR, \fB\-\-auto_tie_breaker\fR, \fB\-\-last_man_standing\fR, \fB\-\-last_man_standing_window\fR options are all documented in corosync's votequorum(5) man page.

\fB\-\-ipv6\fR will configure corosync to use ipv6 (instead of ipv4)

\fB\-\-token\fR <timeout> sets time in milliseconds until a token loss is declared after not receiving a token (default 1000 ms)

\fB\-\-token_coefficient\fR <timeout> sets time in milliseconds used for clusters with at least 3 nodes as a coefficient for real token timeout calculation (token + (number_of_nodes - 2) * token_coefficient) (default 650 ms)

\fB\-\-join\fR <timeout> sets time in milliseconds to wait for join messages (default 50 ms)

\fB\-\-consensus\fR <timeout> sets time in milliseconds to wait for consensus to be achieved before starting a new round of membership configuration (default 1200 ms)

\fB\-\-miss_count_const\fR <count> sets the maximum number of times on receipt of a token a message is checked for retransmission before a retransmission occurs (default 5 messages)

\fB\-\-fail_recv_const\fR <failures> specifies how many rotations of the token without receiving any messages when messages should be received may occur before a new configuration is formed (default 2500 failures)


Configuring Redundant Ring Protocol (RRP)

When using udpu (the default) specifying nodes, specify the ring 0
address first followed by a ',' and then the ring 1 address.

Example: pcs cluster setup \-\-name cname nodeA-0,nodeA-1 nodeB-0,nodeB-1

When using udp, using \fB\-\-addr0\fR and \fB\-\-addr1\fR will allow you to configure
rrp mode for corosync.  It's recommended to use a network (instead of
IP address) for \fB\-\-addr0\fR and \fB\-\-addr1\fR so the same corosync.conf file can
be used around the cluster.  \fB\-\-mcast0\fR defaults to 239.255.1.1 and
\fB\-\-mcast1\fR defaults to 239.255.2.1, \fB\-\-mcastport0/1\fR default to 5405 and
ttl defaults to 1. If \fB\-\-broadcast\fR is specified, \fB\-\-mcast0/1\fR,
\fB\-\-mcastport0/1\fR & \fB\-\-ttl0/1\fR are ignored.
.TP
start [\fB\-\-all\fR] [node] [...]
Start corosync & pacemaker on specified node(s), if a node is not specified then corosync & pacemaker are started on the local node. If \fB\-\-all\fR is specified then corosync & pacemaker are started on all nodes.
.TP
stop [\fB\-\-all\fR] [node] [...]
Stop corosync & pacemaker on specified node(s), if a node is not specified then corosync & pacemaker are stopped on the local node. If \fB\-\-all\fR is specified then corosync & pacemaker are stopped on all nodes.
.TP
kill
Force corosync and pacemaker daemons to stop on the local node (performs kill \-9).
.TP
enable [\fB\-\-all\fR] [node] [...]
Configure corosync & pacemaker to run on node boot on specified node(s), if node is not specified then corosync & pacemaker are enabled on the local node. If \fB\-\-all\fR is specified then corosync & pacemaker are enabled on all nodes.
.TP
disable [\fB\-\-all\fR] [node] [...]
Configure corosync & pacemaker to not run on node boot on specified node(s), if node is not specified then corosync & pacemaker are disabled on the local node. If \fB\-\-all\fR is specified then corosync & pacemaker are disabled on all nodes. (Note: this is the default after installation)
.TP
standby [<node>] | \fB\-\-all\fR
Put specified node into standby mode (the node specified will no longer be able to host resources), if no node or options are specified the current node will be put into standby mode, if \fB\-\-all\fR is specified all nodes will be put into standby mode.
.TP
unstandby [<node>] | \fB\-\-all\fR
Remove node from standby mode (the node specified will now be able to host resources), if no node or options are specified the current node will be removed from standby mode, if \fB\-\-all\fR is specified all nodes will be removed from standby mode.
.TP
remote-node add <hostname> <resource id> [options]
Enables the specified resource as a remote-node resource on the specified hostname (hostname should be the same as 'uname -n')
.TP
remote\-node remove <hostname>
Disables any resources configured to be remote\-node resource on the specified hostname (hostname should be the same as 'uname -n')
.TP
status
View current cluster status (an alias of 'pcs status cluster')
.TP
pcsd\-status [node] [...]
Get current status of pcsd on nodes specified, or on all nodes configured in corosync.conf if no nodes are specified
.TP
certkey <certificate file> <key file>
Load custom certificate and key files for use in pcsd
.TP
sync
Sync corosync configuration to all nodes found from current corosync.conf file (cluster.conf on RHEL6)
.TP
quorum unblock
Cancel waiting for all nodes when establishing quorum.  Usefull in situations where you know the cluster is inquorate, but you are confident that the cluster should proceed with resource management regardless.
.TP
cib [filename] [scope=<scope>]
Get the raw xml from the CIB (Cluster Information Base).  If a filename is provided, we save the cib to that file, otherwise the cib is printed.  Specify scope to get a specific section of the CIB.  Valid values of the scope are: nodes, resources, constraints, crm_config, rsc_defaults, op_defaults, status.
.TP
cib-push <filename> [scope=<scope>]
Push the raw xml from <filename> to the CIB (Cluster Information Base).  Specify scope to push a specific section of the CIB.  Valid values of the scope are: nodes, resources, constraints, crm_config, rsc_defaults, op_defaults.
.TP
cib\-upgrade
Upgrade the cib to the latest version
.TP
edit [scope=<scope>]
Edit the cib in the editor specified by the $EDITOR environment variable and push out any changes upon saving.  Specify scope to edit a specific section of the CIB.  Valid values of the scope are: nodes, resources, constraints, crm_config, rsc_defaults, op_defaults.
.TP
node add <node[,node\-altaddr]> [\fB\-\-start\fR] [\fB\-\-enable\fR]
Add the node to corosync.conf and corosync on all nodes in the cluster and sync the new corosync.conf to the new node.  If \fB\-\-start\fR is specified also start corosync/pacemaker on the new node, if \fB\-\-enable\fR is specified enable corosync/pacemaker on new node.  When using Redundant Ring Protocol (RRP) with udpu transport, specify the ring 0 address first followed by a ',' and then the ring 1 address.
.TP
node remove <node>
Shutdown specified node and remove it from pacemaker and corosync on all other nodes in the cluster
.TP
uidgid
List the current configured uids and gids of users allowed to connect to corosync
.TP
uidgid add [uid=<uid>] [gid=<gid>]
Add the specified uid and/or gid to the list of users/groups allowed to connect to corosync
.TP
uidgid rm [uid=<uid>] [gid=<gid>]
Remove the specified uid and/or gid from the list of users/groups allowed to connect to corosync
.TP
corosync [node]
Get the corosync.conf from the specified node or from the current node if node not specified
.TP
reload corosync
Reload the corosync configuration on the current node
.TP
destroy [\fB\-\-all\fR]
Permanently destroy the cluster on the current node, killing all corosync/pacemaker processes removing all cib files and the corosync.conf file.  Using \fB\-\-all\fR will attempt to destroy the cluster on all nodes configure in the corosync.conf file.  WARNING: This command permantly removes any cluster configuration that has been created. It is recommended to run 'pcs cluster stop' before destroying the cluster.
.TP
verify [\fB\-V\fR] [filename]
Checks the pacemaker configuration (cib) for syntax and common conceptual errors.  If no filename is specified the check is performmed on the currently running cluster.  If \fB\-V\fR is used more verbose output will be printed
.TP
report [\fB\-\-from\fR "YYYY\-M\-D H:M:S" [\fB\-\-to\fR "YYYY\-M\-D" H:M:S"]] dest
Create a tarball containing everything needed when reporting cluster problems.  If \fB\-\-from\fR and \fB\-\-to\fR are not used, the report will include the past 24 hours
.SS "stonith"
.TP
show [stonith id] [\fB\-\-full\fR]
Show all currently configured stonith devices or if a stonith id is specified show the options for the configured stonith device.  If \fB\-\-full\fR is specified all configured stonith options will be displayed
.TP
list [filter] [\fB\-\-nodesc\fR]
Show list of all available stonith agents (if filter is provided then only stonith agents matching the filter will be shown). If \fB\-\-nodesc\fR is used then descriptions of stontih agents are not printed.
.TP
describe <stonith agent>
Show options for specified stonith agent
.TP
create <stonith id> <stonith device type> [stonith device options]
Create stonith device with specified type and options
.TP
update <stonith id> [stonith device options]
Add/Change options to specified stonith id
.TP
delete <stonith id>
Remove stonith id from configuration
.TP
cleanup [<stonith id>]

pcsd/pcs.rb
# Wrapper for PCS command
#
require 'open4'
require 'shellwords'

def getAllSettings()
  stdout, stderr, retval = run_cmd(PCS, "property")
  stdout.map(&:chomp!)
  stdout.map(&:strip!)
  stdout2, stderr2, retval2 = run_cmd(PENGINE, "metadata")
  metadata = stdout2.join
  ret = {}
  if retval == 0 and retval2 == 0
    doc = REXML::Document.new(metadata)

    default = ""
    el_type = ""
    doc.elements.each("resource-agent/parameters/parameter") { |e|
      name = e.attributes["name"]
      name.gsub!(/-/,"_")
      e.elements.each("content") { |c|
	default = c.attributes["default"]
	el_type = c.attributes["type"]
      }
      ret[name] = {"value" => default, "type" => el_type}
    }

    stdout.each {|line|
      key,val = line.split(': ', 2)
      key.gsub!(/-/,"_")
      if ret.has_key?(key)
	if ret[key]["type"] == "boolean"
	  val == "true" ?  ret[key]["value"] = true : ret[key]["value"] = false
	else
	  ret[key]["value"] = val
	end

      else
	ret[key] = {"value" => val, "type" => "unknown"}
      end
    }
    return ret
  end
  return {"error" => "Unable to get configuration settings"}
end

def add_fence_level (level, devices, node, remove = false)
  if not remove
    stdout, stderr, retval = run_cmd(PCS, "stonith", "level", "add", level, node, devices)
    return retval
  else
    stdout, stderr, retval = run_cmd(PCS, "stonith", "level", "remove", level, node, devices)
    return retval
  end
end

def add_node_attr(node, key, value)
  stdout, stderr, retval = run_cmd(PCS, "property", "set", "--node", node, key.to_s + '=' + value.to_s)
  return retval
end

def add_meta_attr(resource, key, value)
  stdout, stderr, retval = run_cmd(PCS,"resource","meta",resource,key.to_s + "=" + value.to_s)
  return retval
end

def add_location_constraint(resource, node, score)
  if node == ""
    return "Bad node"
  end

  if score == ""
    nodescore = node
  else
    nodescore = node +"="+score
  end

  stdout, stderr, retval = run_cmd(
    PCS, "constraint", "location", resource, "prefers", nodescore,
    "--autocorrect"
  )
  return retval, stderr.join(' ')
end

def add_location_constraint_rule(resource, rule, score, force=false)
  cmd = [PCS, "constraint", "location", "--autocorrect", resource, "rule"]
  cmd << "score=#{score}" if score != ""
  cmd.concat(rule.shellsplit())
  cmd << '--force' if force
  stdout, stderr, retval = run_cmd(*cmd)
  return retval, stderr.join(' ')
end

def add_order_constraint(
    resourceA, resourceB, actionA, actionB, score, symmetrical=true, force=false
)
  sym = symmetrical ? "symmetrical" : "nonsymmetrical"
  if score != ""
    score = "score=" + score
  end
  command = [
    PCS, "constraint", "order", actionA, resourceA, "then", actionB, resourceB,
    score, sym, "--autocorrect"
  ]
  command << '--force' if force
  $logger.info command
  Open3.popen3(*command) { |stdin, stdout, stderror, waitth|
    $logger.info stdout.readlines()
    return waitth.value, stderror.readlines().join(' ')
  }
end

def add_order_set_constraint(resource_set_list, force=false)
  command = [PCS, "constraint", "order", "--autocorrect"]
  resource_set_list.each { |resource_set|
    command << "set"
    command.concat(resource_set)
  }
  command << '--force' if force
  stdout, stderr, retval = run_cmd(*command)
  return retval, stderr.join(' ')
end

def add_colocation_constraint(resourceA, resourceB, score, force=false)
  if score == "" or score == nil
    score = "INFINITY"
  end
  command = [
    PCS, "constraint", "colocation", "add", resourceA, resourceB, score,
    "--autocorrect"
  ]
  command << '--force' if force
  $logger.info command
  Open3.popen3(*command) { |stdin, stdout, stderror, waitth|
    $logger.info stdout.readlines()
    return waitth.value, stderror.readlines().join(' ')
  }
end

def remove_constraint(constraint_id)
  stdout, stderror, retval = run_cmd(PCS, "constraint", "remove", constraint_id)
  $logger.info stdout
  return retval
end

def remove_constraint_rule(rule_id)
  stdout, stderror, retval = run_cmd(
    PCS, "constraint", "rule", "remove", rule_id
  )
  $logger.info stdout
  return retval
end

def add_acl_role(name, description)
  cmd = [PCS, "acl", "role", "create", name.to_s]
  if description.to_s != ""
    cmd << "description=#{description.to_s}"
  end
  stdout, stderror, retval = run_cmd(*cmd)
  if retval != 0
    return stderror.join("\n").strip
  end
  return ""
end

def add_acl_permission(acl_role_id, perm_type, xpath_id, query_id)
  stdout, stderror, retval = run_cmd(
    PCS, "acl", "permission", "add", acl_role_id.to_s, perm_type.to_s,
    xpath_id.to_s, query_id.to_s
  )
  if retval != 0
    return stderror.join("\n").strip
  end
  return ""
end

def add_acl_usergroup(acl_role_id, user_group, name)
  if (user_group == "user") or (user_group == "group")
    stdout, stderr, retval = run_cmd(
      PCS, "acl", user_group, "create", name.to_s, acl_role_id.to_s
    )
    if retval == 0
      return ""
    end
    if stderr.join("\n").strip.downcase != "error: #{name.to_s.downcase} already exists in cib"
      return stderror.join("\n").strip
    end
  end
  stdout, stderror, retval = run_cmd(
    PCS, "acl", "role", "assign", acl_role_id.to_s, name.to_s
  )
  if retval != 0
    return stderror.join("\n").strip
  end
  return ""
end

def remove_acl_permission(acl_perm_id)
  stdout, stderror, retval = run_cmd(PCS, "acl", "permission", "delete", acl_perm_id.to_s)
  if retval != 0
    return stderror.join("\n").chomp
  end
  return ""
end

def remove_acl_usergroup(role_id, usergroup_id)
  stdout, stderror, retval = run_cmd(PCS, "acl", "role", "unassign", role_id.to_s, usergroup_id.to_s)
  if retval != 0
    return stderror.join("\n").chomp
  end
  return ""
end

def get_node_token(node)
  out, stderror, retval = run_cmd(PCS, "cluster", "token", node)
  return retval, out
end

def get_token_node_list()
  out, stderror, retval = run_cmd(PCS, "cluster", "token-nodes")
  return retval, out
end

# Gets all of the nodes specified in the pcs config file for the cluster
def get_cluster_nodes(cluster_name)
  pcs_config = PCSConfig.new
  clusters = pcs_config.clusters
  cluster = nil
  for c in clusters
    if c.name == cluster_name
      cluster = c
      break
    end
  end

  if cluster && cluster.nodes != nil
    nodes = cluster.nodes
  else
    $logger.info "Error: no nodes found for #{cluster_name}"
    nodes = []
  end
  return nodes
end

def send_cluster_request_with_token(cluster_name, request, post=false, data={}, remote=true, raw_data=nil)
  out = ""
  code = 0
  nodes = get_cluster_nodes(cluster_name)

  # If we're removing nodes, we don't send this to one of the nodes we're
  # removing, unless we're removing all nodes

  $logger.info("SCRWT: " + request)
  if request == "/remove_nodes"
    new_nodes = nodes.dup
    data.each {|k,v|
      if new_nodes.include? v
        new_nodes.delete v
      end
    }
    if new_nodes.length > 0
      nodes = new_nodes
    end
  end
  for node in nodes
    code, out = send_request_with_token(node,request, post, data, remote=true, raw_data)
    $logger.info "Node: #{node} Request: #{request}"
    if out != '{"noresponse":true}' and out != '{"pacemaker_not_running":true}'
      break
    end
    $logger.info "No response: Node: #{node} Request: #{request}"
  end
  return code,out
end

def send_request_with_token(node,request, post=false, data={}, remote=true, raw_data = nil)
  start = Time.now
  begin
    retval, token = get_node_token(node)
    if retval != 0
      return 400,'{"notoken":true}'
    end

    token = token[0].strip
    if remote
      uri = URI.parse("https://#{node}:2224/remote/" + request)
    else
      uri = URI.parse("https://#{node}:2224/" + request)
    end

    if post
      req = Net::HTTP::Post.new(uri.path)
      raw_data ? req.body = raw_data : req.set_form_data(data)
    else
      req = Net::HTTP::Get.new(uri.path)
      req.set_form_data(data)
    end
    $logger.info("Request: " + uri.to_s + " (" + (Time.now-start).to_s + "s)")
    req.add_field("Cookie","token="+token)
    req.add_field("Cookie","CIB_user="+$session[:username])
    myhttp = Net::HTTP.new(uri.host, uri.port)
    myhttp.use_ssl = true
    myhttp.verify_mode = OpenSSL::SSL::VERIFY_NONE
    res = myhttp.start do |http|
      http.read_timeout = 30 
      http.request(req)
    end
    return res.code.to_i, res.body
  rescue Exception => e
    $logger.info "No response from: #{node} request: #{request}"
    return 400,'{"noresponse":true}'
  end
end

def add_node(new_nodename,all = false, auto_start=true)
  if all
    if auto_start
      out, stderror, retval = run_cmd(PCS, "cluster", "node", "add", new_nodename, "--start", "--enable")
    else
      out, stderror, retval = run_cmd(PCS, "cluster", "node", "add", new_nodename)
    end
  else
    out, stderror, retval = run_cmd(PCS, "cluster", "localnode", "add", new_nodename)
  end
  $logger.info("Adding #{new_nodename} from pcs_settings.conf")
  pcs_config = PCSConfig.new
  pcs_config.update($cluster_name,get_corosync_nodes())
  return retval, out.join("\n") + stderror.join("\n")
end

def remove_node(new_nodename, all = false)
  if all
    out, stderror, retval = run_cmd(PCS, "cluster", "node", "remove", new_nodename)
  else
    out, stderror, retval = run_cmd(PCS, "cluster", "localnode", "remove", new_nodename)
  end
  $logger.info("Removing #{new_nodename} from pcs_settings.conf")
  pcs_config = PCSConfig.new
  pcs_config.update($cluster_name,get_corosync_nodes())
  return retval, out + stderror
end

def get_corosync_nodes()
  stdout, stderror, retval = run_cmd(PCS, "status", "nodes", "corosync")
  if retval != 0
    return []
  end

  stdout.each {|x| x.strip!}
  corosync_online = stdout[1].sub(/^.*Online:/,"").strip
  corosync_offline = stdout[2].sub(/^.*Offline:/,"").strip
  corosync_nodes = (corosync_online.split(/ /)) + (corosync_offline.split(/ /))

  return corosync_nodes
end

# Get pacemaker nodes, but if they are not present fall back to corosync
def get_nodes()
  stdout, stderr, retval = run_cmd(PCS, "status", "nodes")
  if retval != 0
    stdout, stderr, retval = run_cmd(PCS, "status", "nodes", "corosync")
  end

  online = stdout[1]
  offline = stdout[2]

  if online
    online = online.split(' ')[1..-1].sort
  else
    online = []
  end

  if offline
    offline = offline.split(' ')[1..-1].sort
  else
    offline = []
  end

  [online, offline]
end

def get_resource_agents_avail()
  code, result = send_cluster_request_with_token(params[:cluster], 'get_avail_resource_agents')
  ra = JSON.parse(result)
  if (ra["noresponse"] == true) or (ra["notauthorized"] == "true") or (ra["notoken"] == true) or (ra["pacemaker_not_running"] == true)
    return {}
  else
    return ra
  end
end

def get_stonith_agents_avail()
  code, result = send_cluster_request_with_token(params[:cluster], 'get_avail_fence_agents')
  sa = JSON.parse(result)
  if (sa["noresponse"] == true) or (sa["notauthorized"] == "true") or (sa["notoken"] == true) or (sa["pacemaker_not_running"] == true)
    return {}
  else
    return sa
  end
end

def get_cluster_version()
  stdout, stderror, retval = run_cmd(COROSYNC_CMAPCTL,"totem.cluster_name")
  if retval != 0 and not ISRHEL6
    # Cluster probably isn't running, try to get cluster name from
    # corosync.conf
    begin
      corosync_conf = File.open("/etc/corosync/corosync.conf").read
    rescue
      return ""
    end
    in_totem = false
    current_level = 0
    corosync_conf.each_line do |line|
      if line =~ /totem\s*\{/
        in_totem = true
      end
      if in_totem
        md = /cluster_name:\s*(\w+)/.match(line)
        if md
          return md[1]
        end
      end
      if in_totem and line =~ /\}/
        in_totem = false
      end
    end

    return ""
  else
    return stdout.join().gsub(/.*= /,"").strip
  end
end

def get_node_attributes()
  stdout, stderr, retval = run_cmd(PCS, "property", "list")
  if retval != 0
    return {}
  end

  attrs = {}
  found = false
  stdout.each { |line|
    if not found
      if line.strip.start_with?("Node Attributes:")
        found = true
      end
      next
    end
    if not line.start_with?(" ")
      break
    end
    sline = line.split(":", 2)
    nodename = sline[0].strip
    attrs[nodename] = []
    sline[1].strip.split(" ").each { |attr|
      key, val = attr.split("=", 2)
      attrs[nodename] << {:key => key, :value => val}
    }
  }
  return attrs
end

def get_fence_levels()
  stdout, stderr, retval = run_cmd(PCS, "stonith", "level")
  if retval != 0 or stdout == ""
    return {}
  end

  fence_levels = {}
  node = ""
  stdout.each {|line|
    if line.start_with?(" Node: ")
      node = line.split(":",2)[1].strip
      next
    end
    fence_levels[node] ||= []
    md = / Level (\S+) - (.*)$/.match(line)
    fence_levels[node] << {"level" => md[1], "devices" => md[2]}
  }
  return fence_levels
end

def get_acls()
  stdout, stderr, retval = run_cmd(PCS, "acl", "show")
  if retval != 0 or stdout == ""
    return []
  end

  ret_val = {}
  state = nil
  user = ""
  role = ""

  stdout.each do |line|
    if m = /^User: (.*)$/.match(line)
      user = m[1]
      state = "user"
      ret_val[state] ||= {}
      ret_val[state][user] ||= []
      next
    elsif m = /^Group: (.*)$/.match(line)
      user = m[1]
      state = "group"
      ret_val[state] ||= {}
      ret_val[state][user] ||= []
      next
    elsif m = /^Role: (.*)$/.match(line)
      role = m[1]
      state = "role"
      ret_val[state] ||= {}
      ret_val[state][role] ||= {}
      next
    end

    case state
    when "user", "group"
      m = /^  Roles: (.*)$/.match(line)
      ret_val[state][user] ||= []
      m[1].scan(/\S+/).each {|urole|
        ret_val[state][user] << urole
      }
    when "role"
      ret_val[state][role] ||= {}
      ret_val[state][role]["permissions"] ||= []
      ret_val[state][role]["description"] ||= ""
      if m = /^  Description: (.*)$/.match(line)
        ret_val[state][role]["description"] = m[1]
      elsif m = /^  Permission: (.*)$/.match(line)
        ret_val[state][role]["permissions"] << m[1]
      end
    end
  end
  return ret_val
end

def enable_cluster()
  stdout, stderror, retval = run_cmd(PCS, "cluster", "enable")
  return false if retval != 0
  return true
end

def disable_cluster()
  stdout, stderror, retval = run_cmd(PCS, "cluster", "disable")
  return false if retval != 0
  return true
end

def corosync_running?()
  if not ISRHEL6
    `systemctl status corosync.service`
  else
    `service corosync status`
  end
  return $?.success?
end

def corosync_enabled?()
  if not ISRHEL6
    `systemctl is-enabled corosync.service`
  else
    `chkconfig corosync`
  end
  return $?.success?
end

def pacemaker_running?()
  if not ISRHEL6
    `systemctl status pacemaker.service`
  else
    `service pacemaker status`
  end
  return $?.success?
end

def pacemaker_enabled?()
  if not ISRHEL6
    `systemctl is-enabled pacemaker.service`
  else
    `chkconfig pacemaker`
  end
  return $?.success?
end

def cman_running?()
  if not ISRHEL6
    `systemctl status cman.service`
  else
    `service cman status`
  end
  return $?.success?
end

def pcsd_enabled?()
  if not ISRHEL6
    `systemctl is-enabled pcsd.service`
  else
    `chkconfig pcsd`
  end
  return $?.success?
end

def run_cmd(*args)
  $logger.info("Running: " + args.join(" "))
  start = Time.now
  out = ""
  errout = ""
  if $session[:username] == "hacluster"
    ENV['CIB_user'] = $cookies[:CIB_user]
  else
    ENV['CIB_user'] = $session[:username]
  end
  status = Open4::popen4(*args) do |pid, stdin, stdout, stderr|
    out = stdout.readlines()
    errout = stderr.readlines()
    duration = Time.now - start
    $logger.debug(out)
    $logger.debug("Duration: " + duration.to_s + "s")
  end
  retval = status.exitstatus
  $logger.debug("Return Value: " + retval.to_s)
  return out, errout, retval
end
